crates/xtask-lint-docs/src/main.rs

```diff
@@ -0,0 +1,133 @@
+use cargo::util::command_prelude::{flag, ArgMatchesExt};
+use cargo::util::lints::LintLevel;
+use itertools::Itertools;
+use std::fmt::Write;
+use std::path::PathBuf;
+
+struct LintSection {
+    level: LintLevel,
+    names: Vec<&'static str>,
+    docs: String,
+}
+
+impl LintSection {
+    fn new(level: LintLevel) -> Self {
+        Self {
+            level,
+            names: Vec::new(),
+            docs: String::new(),
+        }
+    }
+
+    fn add_lint(&mut self, name: &'static str, docs: &str) {
+        self.names.push(name);
+        self.docs.push_str("### `");
+        self.docs.push_str(name);
+        self.docs.push_str("`\n");
+        self.docs.push_str(docs);
+        self.docs.push_str("\n\n");
+    }
+
+    fn write(&self, buf: &mut String) -> std::fmt::Result {
+        let title = match self.level {
+            LintLevel::Allow => "Allowed-by-default",
+            LintLevel::Warn => "Warn-by-default",
+            LintLevel::Deny => "Deny-by-default",
+            LintLevel::Forbid => "Forbid-by-default",
+        };
+        writeln!(buf, "## {title}\n")?;
+        writeln!(
+            buf,
+            "These lints are all set to the '{}' level by default.",
+            self.level
+        )?;
+        for name in &self.names {
+            writeln!(buf, "- [`{}`](#{})", name, name)?;
+        }
+        writeln!(buf, "\n{}", self.docs)
+    }
+}
+
+fn cli() -> clap::Command {
+    clap::Command::new("xtask-lint-docs").arg(flag("check", "Check that the docs are up-to-date"))
+}
+
+fn main() -> anyhow::Result<()> {
+    let args = cli().get_matches();
+    let check = args.flag("check");
+
+    let file = std::fs::read_to_string("src/cargo/util/lints.rs")?;
+    let mut allow = LintSection::new(LintLevel::Allow);
+    let mut warn = LintSection::new(LintLevel::Warn);
+    let mut deny = LintSection::new(LintLevel::Deny);
+    let mut forbid = LintSection::new(LintLevel::Forbid);
+    for lint in cargo::util::lints::LINTS {
+        let docs = get_lint_docs(lint.name, &file)
```

Comment:
I see no justification for why are we parsing the code for lints.  Its especially odd that we are using cargo-as-a-library to enumerate the lints but then not asking the library for the documentation.

Reply:
I originally used doc comments because that is how you normally comment things in Rust. I switched to having it be a field on `Lint`, and I think it will work fine for now.